src/reflect/value.go

```diff
@@ -644,7 +644,45 @@ func (v Value) MapKeys() []Value {
 	panic("unimplemented: (reflect.Value).MapKeys()")
 }
 
+//go:linkname hashmapStringGet runtime.hashmapStringGetUnsafePointer
+func hashmapStringGet(m unsafe.Pointer, key string, value unsafe.Pointer, valueSize uintptr) bool
+
+//go:linkname hashmapBinaryGet runtime.hashmapBinaryGetUnsafePointer
+func hashmapBinaryGet(m unsafe.Pointer, key, value unsafe.Pointer, valueSize uintptr) bool
+
 func (v Value) MapIndex(key Value) Value {
+	if v.Kind() != Map {
+		panic(&ValueError{Method: "MapIndex", Kind: v.Kind()})
+	}
+
+	// compare key type with actual key type of map
+	if key.typecode != v.typecode.key() {
+		// type error?
+		panic(&ValueError{Method: "MapIndex"})
+	}
+
+	elemType := v.typecode.Elem()
+	elem := New(elemType)
+
+	if key.Kind() == String {
+		if ok := hashmapStringGet(v.pointer(), *(*string)(key.value), elem.value, elemType.Size()); !ok {
+			return Value{}
+		}
+		return elem.Elem()
+	} else if key.typecode.isBinary() {
+		var keyptr unsafe.Pointer
+		if key.isIndirect() {
+			keyptr = key.value
+		} else {
+			keyptr = unsafe.Pointer(&key.value)
+		}
+		//TODO(dgryski): zero out padding bytes in key, if any
```

Comment:
I'm not sure this happens in practice. And if it does, we'd have to make sure to zero the original value I think. For example, this is one place that takes special care to zero the undef parts of a value:

https://github.com/tinygo-org/tinygo/blob/0d56dee00f49bd50eb373c02c30062a75ec28f10/compiler/llvm.go#L82-L87

Reply:
Actually, re-reading the code this does in fact result in undef values, for example if the following value is stored in a struct on a 32-bit architecture:

```go
type small struct {
    x byte
    y int16
}
```

There is some padding, but the `if size < b.targetData.TypeAllocSize(b.i8ptrType) {` check doesn't catch that.

Reply:
Yes, this is exactly the case my recent zero undef padding PR in the compiler fixed.